athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/storage/StorageMetadata.java

```diff
@@ -0,0 +1,286 @@
+            // remove the front-slash, because, the expression generated without it
+            if (folderPath.startsWith("/")) {
+                folderPath = folderPath.substring(1);
+            }
+            LOGGER.info("Examining folder {}", folderPath);
+            List<Map<String, String>> partitions = PartitionUtil.getPartitionColumnData(table, folderPath);
+            if (!partitions.isEmpty() && checkPartitionWithConstrains(partitions, expressions)) {
+                partitionFolders.add(partitions);
+                LOGGER.info("Folder {} is selected", folderPath);
+            }
+        }
+        return partitionFolders;
+    }
+
+    /**
+     * Retrieves a first of file has non-zero size
+     *
+     * @param bucket Name of the bucket
+     * @param prefix Prefix (aka, folder in Storage service) of the bucket from where this method with retrieve files
+     * @return A single file name under the prefix
+     */
+    protected Optional<String> getStorageFiles(String bucket, String prefix)
+    {
+        LOGGER.info("Listing nested files for prefix {} under the bucket {}", prefix, bucket);
+        Page<Blob> blobPage = storage.list(bucket, prefix(prefix.substring(1)));
+        for (Blob blob : blobPage.iterateAll()) {
+            if (isBlobFile(blob)) {
+               return Optional.of(blob.getName());
+            }
+        }
+        return Optional.empty();
+    }
+
+    // helpers
+    /**
+     * check whether it is file, It may return folder also
+     */
+    private boolean isBlobFile(Blob blob)
+    {
+        return blob.getSize() > 0;
+    }
+
+    @VisibleForTesting
+    protected Schema getFileSchema(String bucketName, String path, FileFormat format, BufferAllocator allocator)
+    {
+        requireNonNull(bucketName, "bucketName was null");
+        requireNonNull(path, "fileName was null");
+        LOGGER.info("Retrieving field schema from file {}, under the bucket {}", path, bucketName);
+        String uri = createUri(bucketName, path);
+        DatasetFactory factory = new FileSystemDatasetFactory(allocator,
+                NativeMemoryPool.getDefault(), format, uri);
+        // inspect schema
+        return factory.inspect();
+    }
+
+    private boolean checkPartitionWithConstrains(List<Map<String, String>> partitionList, List<AbstractExpression> expressions)
+    {
+        if (expressions.isEmpty()) {
+            return true;
+        }
+        Map<String, List<AbstractExpression>> expressionMap = expressions.stream()
+                .collect(Collectors.groupingBy(AbstractExpression::getColumnName,
+                        () -> new TreeMap<>(String.CASE_INSENSITIVE_ORDER), toCollection(ArrayList::new)));
+
+        boolean expressionFailureExists = partitionList.stream().anyMatch(partitionEntry ->
+                (expressionMap.getOrDefault(partitionEntry.keySet().stream().findFirst().get(), List.of()).stream()
```

Comment:
Why are we doing a findFirst() here? aren't we trying to validate the entire partitionList? Otherwise why is a list being passed in here at all?